rb_link/src/server.rs

```diff
@@ -0,0 +1,284 @@
+            let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+            if let Some(queue) = b2r_cache.get_mut(&id) {
+                if let Some(b2r_message) = queue.pop_front() {
+                    return b2r_message;
+                }
+            }
+            drop(b2r_cache);
+            thread::sleep(Duration::from_millis(100));
+        }
+    }
+
+    /// Return the earliest message from the probe with id.
+    /// This function will return None if there is no message available for retrieval.
+    pub fn try_get(&self, id: u32) -> Option<B2RMessage> {
+        let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+        if let Some(queue) = b2r_cache.get_mut(&id) {
+            if let Some(b2r_message) = queue.pop_front() {
+                return Some(b2r_message);
+            }
+        }
+        None
+    }
+
+    /// Send a message to the probe with ID "id".
+    /// Please ensure that message.len() == ceil(get_t_width/8),
+    /// where get_t_width is the width of get_t defined in your BSV code.
+    pub fn put(&mut self, id: u32, message: Vec<u8>) {
+        let r2b_message = R2BMessage { id, message };
+        let mut r2b_cache = self.r2b_cache.lock().expect("Fail to lock r2b_cache");
+        let queue = r2b_cache.entry(id).or_default();
+        queue.push_back(r2b_message);
+    }
+
+    /// Get all the messages sent by the earliest cycle.
+    /// If there are no messages available, it will return an empty Vec.
+    pub fn get_cycle_message(&mut self) -> Vec<B2RMessage> {
+        let mut min_cycles = u32::MAX;
+        let mut messages: Vec<B2RMessage> = Vec::new();
+        let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+        for queue in b2r_cache.values() {
+            if let Some(b2r_message) = queue.front() {
+                if b2r_message.cycles < min_cycles {
+                    min_cycles = b2r_message.cycles;
+                }
+            }
+        }
+        for queue in b2r_cache.values_mut() {
+            if let Some(b2r_message) = queue.front() {
+                if b2r_message.cycles == min_cycles {
+                    messages.push(queue.pop_front().expect("front error"));
+                }
+            }
+        }
+        messages
+    }
+
+    /// Get all message send by the probe with id.
+    pub fn get_id_all(&mut self, id: u32) -> Vec<B2RMessage> {
+        let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+        let mut messages: Vec<B2RMessage> = Vec::new();
+
+        if let Some(queue) = b2r_cache.get_mut(&id) {
+            while let Some(b2r_message) = queue.pop_front() {
+                messages.push(b2r_message);
+            }
+        }
+        messages
+    }
+
+    /// Read the earliest cycle messages sent by the probes labeled as "fifo" and "fired", and organize them into a PipeLineState.
+    pub fn get_pipeline_state(&mut self) -> PipeLineState {
```

Comment:
What's the purpose of the Server module? From the name, we think that it should be a module that handles connections, e.g., accept connection, retry connection. But what does the server do after receive a packet should not be included in a module called Server. And maybe there will be many different processing logics, should we put all thoes new logics into the Server module? This is a bad smell in code. Use some design pattern to refactor this.

Reply:
fixed